src/JsonRpc/InputHandler.cs

```diff
@@ -39,113 +58,234 @@ public class InputHandler : IInputHandler
             ILoggerFactory loggerFactory,
             ISerializer serializer,
             int? concurrency
-            )
+        )
         {
-            if (!input.CanRead) throw new ArgumentException($"must provide a readable stream for {nameof(input)}", nameof(input));
-            _input = input;
+            _pipeReader = pipeReader;
             _outputHandler = outputHandler;
             _receiver = receiver;
             _requestProcessIdentifier = requestProcessIdentifier;
             _requestRouter = requestRouter;
             _responseRouter = responseRouter;
             _serializer = serializer;
             _logger = loggerFactory.CreateLogger<InputHandler>();
-            _scheduler = new ProcessScheduler(loggerFactory, concurrency);
-            _inputThread = new Thread(ProcessInputStream) { IsBackground = true, Name = "ProcessInputStream" };
+            _scheduler = new ProcessScheduler(loggerFactory, concurrency,
+                new EventLoopScheduler(_ => new Thread(_) {IsBackground = true, Name = "InputHandler"}));
+            _headersBuffer = new Memory<byte>(new byte[HeadersFinishedLength]);
+            _contentLengthBuffer = new Memory<byte>(new byte[ContentLengthLength]);
+            _contentLengthValueBuffer = new byte[20]; // Max string length of the long value
+            _contentLengthValueMemory =
+                new Memory<byte>(_contentLengthValueBuffer); // Max string length of the long value
+            _stopProcessing = new CancellationTokenSource();
+
+            _disposable = new CompositeDisposable {
+                Disposable.Create(() => _stopProcessing.Cancel()),
+                _stopProcessing,
+                _scheduler,
+            };
         }
 
         public void Start()
         {
-            _scheduler.Start();
-            _outputHandler.Start();
-            _inputThread.Start();
+            ProcessInputStream(_stopProcessing.Token).ContinueWith(x => {
+                if (x.IsFaulted) _logger.LogCritical(x.Exception, "unhandled exception");
+            });
         }
 
-        // don't be async: We already allocated a seperate thread for this.
-        private void ProcessInputStream()
+        private bool TryParseHeaders(ref ReadOnlySequence<byte> buffer, out ReadOnlySequence<byte> line)
```

Comment:
The parsing logic in here can be simplified using `SequenceReader<byte>`, it's currently only in .NET Core 3.0 but Nerdbank.Streams has https://github.com/AArnott/Nerdbank.Streams/blob/master/doc/SequenceTextReader.md that you should be able to use to simplify this logic.

It's slightly allocatey though, it might make sense to copy the SequenceReader sources into this ns2.0 project from .NET Core.